### ISSUE TITLE
BN and Scale layer ?

### ISSUE THREAD

**Issue by ananddb90** [kai-xie/Neural-Network-Model-Compressing#2]
Hi @kai-xie 
Thank you for great work and share.
I would like to use your work for detection network using resnet architecture. Since it consists of conv-bn-scale layers only (No InnerProduct layer).
1. After network pretraining, I have to run pruning for conv layer with itr_size = max_itr
2. What about BN and scale layer ? Is there is any parameter to take care for it
3. While pruning, do I have to make my base learning rate small or same as normal training ?
4. could you please give me some information about usage of scripts available under compression_scripts/
5. which branch do I have to use for dynamic_pruning : master or dev ?
6. Does this branch include incremental quantization also ?

-anand

**Comment by kai-xie** [comment]
Hi @ananddb90 
To be honest, this repo is not stable yet, files are kinda messy, and I am still working on it. These are the problems I've encountered https://github.com/yiwenguo/Dynamic-Network-Surgery/issues/18 , it does not converge when pruning large neural networks (e.g. AlexNet) with high 'c_rate' (i.e. high compression rate) and I have not figured out the reason. 

So for now, I would not recommend using this repo for your actual work. The code is basically the same as Yiwen Guo's version, but a little buggy when transfering to this BVLC version, and If you are interest in the code or have any suggestions, you are very welcome.

As for your questions:

1. I'am not quite sure what iter_size and max_itr refer to, but the DNS hyperparamtes 'iter_stop' denotes when the pruning stops, it could be either larger or smaller than 'max_iter', if smaller, DNS pruning process would stop before the training process stops.

2. As far as my understanding, BN and scale layer do not matter much in DNS. 

3. I usually just try it when setting the initial learning rate, setting lr as normal first, if it does not converge, try smaller ones.

4. Take the 'dns_to_normal.py' as an example (this script is used for converting the raw DNS caffemodel to normal caffemodel and checking the compression rate for each learnable layer):

After compiling the caffe and pycaffe, prepare your compressed DNS caffemodel, and run the following command from your CAFFE_ROOT (make sure you have set `CAFFE_ROOT` environment variable, which is the dir of you caffe folder) :

``` shell
python compression_scripts/dns_to_normal.py <dns.prototxt> <dns_model.caffemodel> <target.prototxt> <output_target.caffemodel>
```

The compression rate should be shown on the screen, and the `output_target.caffemodel` should have the same size as a normal caffemodel (about 1/2 of the dns_model.caffemodel) which can be tested.

e.g.

``` shell
python compression_scripts/dns_to_normal.py examples/mnist/dns_train_val.prototxt examples/mnist/dns_iter_10000.caffemodel examples/mnist/mnist_train_val.prototxt examples/mnist/mnist_test.caffemodel
```

'dns_to_normal.py' is compatible with Yiwen Guo' version.

The other scripts are just some trial scripts, you can just ignore them.

5. try 'dev'.

6.  INQ is under debug, so I haven't uploaded it yet. INQ's author has also released his source code on github https://github.com/Zhouaojun/Incremental-Network-Quantization. I haven't check the details of it but it seems their code is independent of DNS, so you have to do some work on your own to combine those two algorithms. My version will combine them into one set of code and some more scripts for model conversion and compressing the model to a smaller size. Hopefully this can be done next month.

I would recommend you to check the official DNS and INQ source code and their issues, they are very helpful.

**Comment by ananddb90** [comment]
@kai-xie 
thank you so much for your time and detailed reply. This is the best reply I got so far :dancing_men: 
I will write you if I need couldn't understand something

**Comment by ananddb90** [comment]
@kai-xie 
After pruning the caffemodel, output will be sparse caffemodel. 
1. Since number of zero elements are still stored in the caffemodel so size of the model is same. Do you know what one has to do to reduce the size of caffemodel after pruning
2. Also, does sparse caffemodel will increase inference time ? or we need special library for sparse matrix or different sparse matrix representation should be used in order to reduce inference time ?

Thank you
